app/registries/foreman/setting_manager.rb

```diff
@@ -0,0 +1,74 @@
+module Foreman
+  class SettingManager
+    class << self
+      # Holds all the setting definitions for application
+      def settings
+        @settings ||= {}
+      end
+
+      # Holds all setting categories with their labels
+      def categories
+        @categories ||= { 'general' => N_('General') }
+      end
+
+      def map(context_name, &block)
+        mapper = Mapper.new(self, context_name.to_sym)
+        if block_given?
+          mapper.instance_eval(&block)
+        else
+          mapper
+        end
+      end
+    end
+
+    class Mapper
+      attr_reader :context_name, :current_category
+
+      def initialize(manager, context_name)
+        @manager = manager
+        @context_name = context_name
+        @current_category = 'general'
+      end
+
+      def set_category(category_name, category_label = nil)
+        @manager.categories[category_name.to_s] ||= category_label
+        @current_category = category_name.to_s
+      end
+
+      def category(category_name, category_label = nil, &block)
+        prev_category = @current_category
+        set_category(category_name, category_label)
+        instance_eval(&block)
+      ensure
+        @current_category = prev_category
+      end
+
+      # Adds setting definition
+      #
+      # ===== Example
+      #
+      #   SettingManager.map(:puppet) do
+      #     category(:cfgmgmt, N_('Configuration Management')) do
+      #       setting(:use_cooler_puppet, default: true,
+      #                                   description: N_('This will ensure we install a new and shiny Puppet on hosts instead of a rusty one'),
+      #                                   full_name: N_('Use shiny puppet'),
+      #                                   encrypt: true)
+      #     end
+      #   end
+      #
+      def setting(name, default:, description:, full_name: nil, value: nil, collection: nil, encrypted: false, **options)
+        @manager.settings[name.to_s] = {
```

Comment:
Should we make it so simple to override setting?
Or should we refuse to override, and have special method `override`?

Reply:
IMHO overriding settings is never a good idea. It will depend no load orders and can only lead to problems IMHO. A hard exception without an override is where I'd start.